include/xtensor/xmathutil.hpp

```diff
@@ -0,0 +1,423 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay and Wolf Vollprecht    *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+/**
+ * @brief standard mathematical functions for xexpressions
+ */
+
+#ifndef XMATHUTIL_HPP
+#define XMATHUTIL_HPP
+
+#include <cmath>
+#include <cstdlib>   // std::abs(int) prior to C++ 17
+#include <complex>
+#include <algorithm> // std::min, std::max
+#include <type_traits>
+
+#include "xconcepts.hpp"
+
+namespace xt
+{
+    template <class T = double>
+    struct numeric_constants
+    {
+        static constexpr T PI = 3.141592653589793238463;
+        static constexpr T PI_2 = 1.57079632679489661923;
+        static constexpr T PI_4 = 0.785398163397448309616;
+        static constexpr T D_1_PI = 0.318309886183790671538;
+        static constexpr T D_2_PI = 0.636619772367581343076;
+        static constexpr T D_2_SQRTPI = 1.12837916709551257390;
+        static constexpr T SQRT2 = 1.41421356237309504880;
+        static constexpr T SQRT1_2 = 0.707106781186547524401;
+        static constexpr T E = 2.71828182845904523536;
+        static constexpr T LOG2E = 1.44269504088896340736;
+        static constexpr T LOG10E = 0.434294481903251827651;
+        static constexpr T LN2 = 0.693147180559945309417;
+    };
+
+        /** @brief Namespace for the standard algebraic functions.
+
+            Often, one only wants to import the algebraic functions from
+            namespace 'std'. Standard C++ doesn't support this -- one either
+            has to import every function individually, or writes <tt>using namespace std</tt>
+            to import everything. Now you can write <tt>using namespace xt::cmath</tt>.
+
+            This namespace also adds a few overloads to the standard function (abs() for unsigned
+            types, floor() and ceil() for integer types) which are needed to dismbiguate
+            templated code.
+        */
+    namespace cmath
+    {
+
+        using std::abs;
```

Comment:
Note, these are already available in the `xmath` namespace in the form of universal functions.

Reply:
The intention of this namespace is different: I wanted to put the standard algebraic functions into a namespace of their own, so that one can use the idiom
```c++
    using namespace cmath;
    auto x = sqrt(y);
```
to have the compiler perform both argument dependent lookup and lookup in namespace std (if `y` is a scalar). Without namespace cmath, one would have to write `using std::sqrt;` (for every function one wants to use) or `using namespace std;` (possibly importing functions one didn't want to import).